pkg/controller/storagecluster/storagecluster.go

```diff
@@ -0,0 +1,733 @@
+					cluster.Namespace,
+					&podTemplate,
+					cluster,
+					metav1.NewControllerRef(cluster, controllerKind),
+				)
+				if err != nil && errors.IsTimeout(err) {
+					// TODO
+					// Pod is created but its initialization has timed out. If the initialization
+					// is successful eventually, the controller will observe the creation via the
+					// informer. If the initialization fails, or if the pod keeps uninitialized
+					// for a long time, the informer will not receive any update, and the controller
+					// will create a new pod.
+					return
+				}
+				if err != nil {
+					logrus.Warnf("Failed creation of storage pod on node %v: %v", nodesNeedingStoragePods[idx], err)
+					errCh <- err
+				}
+			}(i)
+		}
+		createWait.Wait()
+		skippedPods := createDiff - batchSize
+		if errorCount < len(errCh) && skippedPods > 0 {
+			logrus.Debugf("Slow-start failure. Skipping creation of %d pods", skippedPods)
+			// The skipped pods will be retried later. The next controller resync will
+			// retry the slow start process.
+			break
+		}
+	}
+
+	logrus.Debugf("Pods to delete for storage cluster %s: %+v, deleting %d",
+		cluster.Name, podsToDelete, deleteDiff)
+	deleteWait := sync.WaitGroup{}
+	deleteWait.Add(deleteDiff)
+	for i := 0; i < deleteDiff; i++ {
+		go func(idx int) {
+			defer deleteWait.Done()
+			err := c.podControl.DeletePod(
+				cluster.Namespace,
+				podsToDelete[idx],
+				cluster,
+			)
+			if err != nil {
+				logrus.Warnf("Failed deletion of storage pod %v: %v", podsToDelete[idx], err)
+				errCh <- err
+			}
+		}(i)
+	}
+	deleteWait.Wait()
+
+	errors := []error{}
+	close(errCh)
+	for err := range errCh {
+		errors = append(errors, err)
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+func (c *Controller) podsShouldBeOnNode(
+	node *v1.Node,
+	nodeToStoragePods map[string][]*v1.Pod,
+	cluster *corev1alpha1.StorageCluster,
+) (nodesNeedingStoragePods, podsToDelete []string, err error) {
+	wantToRun, shouldSchedule, shouldContinueRunning, err := c.nodeShouldRunStoragePod(node, cluster)
+	if err != nil {
+		return
+	}
+
+	storagePods, exists := nodeToStoragePods[node.Name]
+
+	switch {
+	case wantToRun && !shouldSchedule:
+		// If storage pod is supposed to run, but cannot be scheduled, add to suspended list.
+		// TODO
+	case shouldSchedule && !exists:
+		nodesNeedingStoragePods = append(nodesNeedingStoragePods, node.Name)
+	case shouldContinueRunning:
+		// If a storage pod failed, delete it.
+		// If there's no storage pod left on this node, we will create it in the next sync loop
+		var storagePodsRunning []*v1.Pod
+		for _, pod := range storagePods {
+			if pod.DeletionTimestamp != nil {
+				continue
+			}
+			if pod.Status.Phase == v1.PodFailed {
+				msg := fmt.Sprintf("Found failed storage pod %s on node %s, will try to kill it", pod.Name, node.Name)
+				logrus.Warnf(msg)
+				c.recorder.Eventf(cluster, v1.EventTypeWarning, FailedStoragePodReason, msg)
+				podsToDelete = append(podsToDelete, pod.Name)
```

Comment:
A pod to replace the one we are deleting will get scheduled in the next sync loop?

Reply:
The pods we are deleting here are the ones that are in failed state. 
Irrespective of that, as the pod is deleted the controller gives us a callback on pod changes. We don't have to wait till the next sync loop as we are watching over pod changes.